vms/platformvm/spend.go

```diff
@@ -24,46 +24,57 @@ import (
-	errUnknownOwners                = errors.New("unknown owners")
+	errWrongLockState               = errors.New("wrong lock state reported")
+	errInvalidTargetLockState       = errors.New("invalid target lock state")
+	errUnknownOwnersType            = errors.New("unknown owners")
 	errCantSign                     = errors.New("can't sign")
+	errInputsCredentialsMismatch    = errors.New("number of inputs is different from number of credentials")
+	errInputsUTXOSMismatch          = errors.New("number of inputs is different from number of utxos")
+	errWrongCredentials             = errors.New("wrong credentials")
+	errLockingLockedUTXO            = errors.New("utxo consumed for locking are already locked")
+	errLockedInsOrOuts              = errors.New("transaction body has locked inputs or outs, but that's now allowed")
+	errWrongProducedAmount          = errors.New("produced more tokens, than input had")
+	errNotBurnedEnough              = errors.New("burned less tokens, than we need to")
 )
 
-// stake the provided amount while deducting the provided fee.
+// spend the provided amount while deducting the provided fee.
 // Arguments:
 // - [keys] are the owners of the funds
-// - [amount] is the amount of funds that are trying to be staked
-// - [fee] is the amount of AVAX that should be burned
+// - [totalAmountToSpend] is the amount of funds that are trying to be spent (changed their state)
+// - [totalAmountToBurn] is the amount of AVAX that should be burned
+// - [appliedLockState] is lockState that will be applied to consumed outs state
 // Returns:
 // - [inputs] the inputs that should be consumed to fund the outputs
-// - [returnedOutputs] the outputs that should be immediately returned to the
-//                     UTXO set
-// - [stakedOutputs] the outputs that should be locked for the duration of the
-//                   staking period
+// - [outputs] the outputs produced as result of spending
 // - [signers] the proof of ownership of the funds being moved
-func (vm *VM) stake(
+func (vm *VM) spend(
 	keys []*crypto.PrivateKeySECP256K1R,
-	amount uint64,
-	fee uint64,
+	totalAmountToSpend uint64,
+	totalAmountToBurn uint64,
+	appliedLockState LockState,
 ) (
 	[]*avax.TransferableInput, // inputs
-	[]*avax.TransferableOutput, // returnedOutputs
-	[]*avax.TransferableOutput, // stakedOutputs
+	[]*avax.TransferableOutput, // outputs
 	[][]*crypto.PrivateKeySECP256K1R, // signers
 	error,
 ) {
+	if appliedLockState != LockStateBonded && appliedLockState != LockStateDeposited {
+		return nil, nil, nil, errInvalidTargetLockState
+	}
+
 	addrs := ids.NewShortSet(len(keys)) // The addresses controlled by [keys]
 	for _, key := range keys {
 		addrs.Add(key.PublicKey().Address())
 	}
 	utxos, err := avax.GetAllUTXOs(vm.internalState, addrs) // The UTXOs controlled by [keys]
 	if err != nil {
-		return nil, nil, nil, nil, fmt.Errorf("couldn't get UTXOs: %w", err)
+		return nil, nil, nil, fmt.Errorf("couldn't get UTXOs: %w", err)
 	}
 
```

Comment:
utxos should be sorted / filtered first
- LockState set: filter
- Locked first (sorted by !Lockstate tXId)
- Unlocked

After that it should be also possible to handle spending in a single UTXO loop instead of 2.
See next "Improvement" comment